hip/test/components/cooperative_groups.hip.cpp

```diff
@@ -262,4 +266,53 @@ TEST_F(CooperativeGroups, SubwarpBallot) { test(cg_subwarp_ballot); }
 TEST_F(CooperativeGroups, SubwarpBallot2) { test_subwarp(cg_subwarp_ballot); }
 
 
+template <typename ValueType>
+__global__ void cg_shuffle_sum(const int num, ValueType *__restrict__ value)
+{
+    auto group =
+        group::tiled_partition<config::warp_size>(group::this_thread_block());
+    for (int ind = 0; ind < num; ind++) {
+        value[group.thread_rank()] += group.shfl(value[ind], ind);
+    }
+}
+
+TEST_F(CooperativeGroups, ShuffleSumDouble)
+{
+    int num = 4;
+    uint64_t x = 0x401022C90008B240;
+    gko::Array<double> value(ref, config::warp_size);
+    gko::Array<double> answer(ref, config::warp_size);
+    gko::Array<double> dvalue(hip);
+    for (int i = 0; i < value.get_num_elems(); i++) {
+        value.get_data()[i] = *reinterpret_cast<double *>(&x);
```

Comment:
Here, I would not use `reinterpret_cast` since it results in *undefined behavior* (see [above *Defect Reports* here](https://en.cppreference.com/w/cpp/language/reinterpret_cast) and the linked [letter to SG12](http://www.open-std.org/pipermail/ub/2016-February/000565.html)). Since we don't care about runtime in this case, I would propose to do it the proper way (might even be better optimized):
```
#include <string> // needed for memcpy
// ...
uint64_t x = 0x401022C90008B240; // Not sure if this is necessary, but assume it is
double x_dbl{};
std::memcpy(&x_dbl, &x, sizeof(x_dbl)); // now, x_dbl contains the bits of x
```
This has 2 advantages:
- We comply with the standard
- You only do it once instead of `config::warp_size`  times (this should definitively change)